src/olympia/translations/tests/test_models.py

```diff
@@ -678,15 +687,120 @@ def test_comparison_with_lazy():
+
+    # Update!
+    translation.activate('en-us')
+    o = TranslatedModelWithDefaultNull.objects.get(pk=o.pk)
+    translation_id = o.name.autoid
+
+    o.name = 'new name'
+    o.save()
+
+    o = TranslatedModelWithDefaultNull.objects.get(pk=o.pk)
+    assert o.name == 'new name'
+    assert o.name.locale == 'en-us'
+    # Make sure it was an update, not an insert.
+    assert o.name.autoid == translation_id
+
+    # Set translations with a dict.
+    strings = {'en-us': 'right language', 'de': 'wrong language'}
+    o = TranslatedModelWithDefaultNull.objects.create(name=strings)
+
+    # Make sure we get the English text since we're in en-US.
+    assert o.name == 'right language'
+
+    # Check that de was set.
+    translation.activate('de')
+    o = TranslatedModelWithDefaultNull.objects.get(pk=o.pk)
+    assert o.name == 'wrong language'
+    assert o.name.locale == 'de'
+
+
+def test_translated_field_fk_lookups():
+    """
+    Test that translations are properly resolved even through models
+    that are one foreign-key layer away
+    (e.g Version -> License -> Translation).
+
+    The problem here was, that we did not set `base_manager_name` on
+    the `ModelBase`. This superseeded setting `use_for_related_fields`.
+    """
+    assert Translation.objects.count() == 0
+    assert TranslatedModelLinkedAsForeignKey.objects.count() == 0
+    obj = TranslatedModelLinkedAsForeignKey.objects.create(name='english name')
+
+    def get_model():
+        return TranslatedModelLinkedAsForeignKey.objects.get(pk=obj.pk)
+
+    assert Translation.objects.count() == 1
+
+    # Make sure the translation id is stored on the model, not the autoid.
+    assert obj.name.id == obj.name_id
+
+    # Reload the object from database with a different locale activated.
+    # Its name should still be there, using the fallback...
+    translation.activate('de')
+    german = get_model()
+    assert german.name == 'english name'
+    assert german.name.locale == 'en-us'
+
+    # Check that a different locale creates a new row with the same id.
+    german.name = u'Gemütlichkeit name'
+    german.save()
+
+    assert Translation.objects.count() == 2  # New name *and* description.
+    assert german.name == u'Gemütlichkeit name'
+    assert german.name.locale == 'de'
+
+    # Now fetch the parent `TranslatedModel` and make sure that
+    # all the relevant translations from `TranslatedModelLinkedAsForeignKey`
+    # are properly loaded.
+    print('-------------------------------------------------------------')
+    parent = TranslatedModel.objects.create(name='parent')
+    parent.translated_through_fk = obj
+    parent.save()
+
+    # This still works, simply attaching the model
+    assert parent.translated_through_fk.name_id == obj.name_id
+    assert parent.translated_through_fk.name is not None
+
+    print('.............................................................')
+    # This fails... I have no idea yet
```

Comment:
heh, I do have an idea how and this comment needs to be removed :rofl: 

We can uplift all of these tests.